cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizerInstantiatorService.java

```diff
@@ -0,0 +1,281 @@
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!authorizationEnabled) {
+      this.authorizer = new NoOpAuthorizer();
+      return;
+    }
+    // Authorization is enabled, so continue with startup now
+    String authorizerExtensionJarPath = cConf.get(Constants.Security.Authorization.AUTHORIZER_EXTENSION_JAR_PATH);
+    if (Strings.isNullOrEmpty(authorizerExtensionJarPath)) {
+      throw new IllegalArgumentException(
+        String.format("Authorizer extension jar path not found in configuration. Please set %s in cdap-site.xml to " +
+                        "the fully qualified path of the jar file to use as the authorization backend.",
+                      Constants.Security.Authorization.AUTHORIZER_EXTENSION_JAR_PATH));
+    }
+    File authorizerExtensionJar = new File(authorizerExtensionJarPath);
+    ensureValidAuthExtensionJar(authorizerExtensionJar);
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    this.tmpDir = DirUtils.createTempDir(tmpDir);
+    this.authorizerClassLoader = createAuthorizerClassLoader(authorizerExtensionJar);
+    this.authorizer = createAuthorizerInstance(authorizerExtensionJar);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (!authorizationEnabled) {
+      // nothing to close, since we would not have created a class loader
+      return;
+    }
+    try {
+      authorizerClassLoader.close();
+    } catch (IOException e) {
+      LOG.warn("Failed to close authorizer class loader", e);
+    }
+    try {
+      DirUtils.deleteDirectoryContents(tmpDir);
+    } catch (IOException e) {
+      // It's a cleanup step. Nothing much can be done if cleanup fails.
+      LOG.warn("Failed to delete directory {}", tmpDir);
+    }
+  }
+
+  /**
+   * @return an instance of the configured {@link Authorizer} extension, or of {@link NoOpAuthorizer}, if authorization
+   * is disabled
+   */
+  public Authorizer get() throws IOException, InvalidAuthorizerException {
+    return authorizer;
+  }
+
+  /**
+   * Creates a new instance of the configured {@link Authorizer} extension, based on the provided extension jar file.
+   *
+   * @return a new instance of the configured {@link Authorizer} extension
+   */
+  public Authorizer createAuthorizerInstance(File authorizerExtensionJar)
+    throws IOException, InvalidAuthorizerException {
+    Class<? extends Authorizer> authorizerClass = loadAuthorizerClass(authorizerExtensionJar);
+    // Set the context class loader to the AuthorizerClassLoader before creating a new instance of the extension,
+    // so all classes required in this process are created from the AuthorizerClassLoader.
+    ClassLoader oldClassLoader = Thread.currentThread().getContextClassLoader();
+    Thread.currentThread().setContextClassLoader(authorizerClassLoader);
```

Comment:
Merge the get and set call into one. The `setContextClassLoader` will return you the current context classloader before the change happened.